File: src/main/java/org/example/db/UserDAO.java
package org.example.db;

import org.example.DTO.User;
import java.sql.*;

public class UserDAO {
    private final Connection conn;

    public UserDAO() {
        conn = DatabaseManager.connect();
    }

    // 회원 등록 (CREATE)
    public boolean insertUser(User user) {
        String sql = "INSERT INTO users (id, password_hash, license_verified) VALUES (?, ?, ?)";
        try (PreparedStatement pstmt = conn.prepareStatement(sql)) {
            pstmt.setString(1, user.getId());
            pstmt.setString(2, user.getPasswordHash());
            pstmt.setInt(3, user.isLicenseVerified() ? 1 : 0);
            pstmt.executeUpdate();
            System.out.println("✅ 회원 등록 완료: " + user.getId());
            return true;
        } catch (SQLException e) {
            System.out.println("❌ 회원 등록 실패: " + e.getMessage());
            return false;
        }
    }

    // 회원 조회 (READ)
    public User getUserById(String id) {
        String sql = "SELECT * FROM users WHERE id = ?";
        try (PreparedStatement pstmt = conn.prepareStatement(sql)) {
            pstmt.setString(1, id);
            ResultSet rs = pstmt.executeQuery();

            if (rs.next()) {
                return User.builder()
                        .id(rs.getString("id"))
                        .passwordHash(rs.getString("password_hash"))
                        .licenseVerified(rs.getInt("license_verified") == 1)
                        .createdAt(rs.getString("created_at"))
                        .build();
            }
        } catch (SQLException e) {
            System.out.println("❌ 회원 조회 실패: " + e.getMessage());
        }
        return null;
    }

    // 3️⃣ 면허 인증 상태 갱신 (UPDATE)
    public boolean updateLicense(String id, boolean verified) {
        String sql = "UPDATE users SET license_verified = ? WHERE id = ?";
        try (PreparedStatement pstmt = conn.prepareStatement(sql)) {
            pstmt.setInt(1, verified ? 1 : 0);
            pstmt.setString(2, id);
            int result = pstmt.executeUpdate();
            if (result > 0) {
                System.out.println("✅ 면허 상태 변경 완료 (" + id + " → " + verified + ")");
                return true;
            }
        } catch (SQLException e) {
            System.out.println("❌ 면허 상태 변경 실패: " + e.getMessage());
        }
        return false;
    }

    // 회원 삭제 (DELETE)
    public boolean deleteUser(String id) {
        String sql = "DELETE FROM users WHERE id = ?";
        try (PreparedStatement pstmt = conn.prepareStatement(sql)) {
            pstmt.setString(1, id);
            int result = pstmt.executeUpdate();
            if (result > 0) {
                System.out.println("✅ 회원 삭제 완료: " + id);
                return true;
            }
        } catch (SQLException e) {
            System.out.println("❌ 회원 삭제 실패: " + e.getMessage());
        }
        return false;
    }
}

File: lib/data/user_dao.dart
import 'package:sqflite/sqflite.dart';
import '../models/user.dart';
import 'database_manager.dart';

class UserDAO {
  final DatabaseManager _dbManager = DatabaseManager();

  Future<Database> get _db async => await _dbManager.database;

  // 로그인 (SELECT with password)
  Future<User?> logIn(String userId, String hashPassword) async {
    String sql = 'SELECT * FROM users WHERE id = ? AND passwordHash = ?';
    try {
      final db = await _db;
      final List<Map<String, dynamic>> maps = await db.rawQuery(
        sql,
        [userId, hashPassword],
      );

      if (maps.isNotEmpty) {
        return User.fromMap(maps.first);
      }
    } catch (e) {
      print('❌ 로그인 쿼리 실행 실패: $e');
    }

    return null; // 로그인 실패
  }

  // 회원 등록 (CREATE)
  Future<bool> insertUser(User user) async {
    String sql = 'INSERT INTO users (id, passwordHash, licenseVerified, createdAt) VALUES (?, ?, ?, ?)';
    try {
      final db = await _db;
      await db.rawInsert(
        sql,
        [
          user.id,
          user.passwordHash,
          user.licenseVerified ? 1 : 0,
          user.createdAt,
        ],
      );
      print('✅ 회원 등록 완료: ${user.id}');
      return true;
    } catch (e) {
      print('❌ 회원 등록 실패: $e');
      return false;
    }
  }

  // 회원 조회 (READ)
  Future<User?> getUserById(String id) async {
    String sql = 'SELECT * FROM users WHERE id = ?';
    try {
      final db = await _db;
      final List<Map<String, dynamic>> maps = await db.rawQuery(sql, [id]);

      if (maps.isNotEmpty) {
        return User.fromMap(maps.first);
      }
    } catch (e) {
      print('❌ 회원 조회 실패: $e');
    }
    return null;
  }

  // 면허 인증 상태 갱신 (UPDATE)
  Future<bool> updateLicense(String id, bool verified) async {
    String sql = 'UPDATE users SET licenseVerified = ? WHERE id = ?';
    try {
      final db = await _db;
      int result = await db.rawUpdate(sql, [verified ? 1 : 0, id]);
      if (result > 0) {
        print('✅ 면허 상태 변경 완료 ($id → $verified)');
        return true;
      }
    } catch (e) {
      print('❌ 면허 상태 변경 실패: $e');
    }
    return false;
  }

  // 회원 삭제 (DELETE)
  Future<bool> deleteUser(String id) async {
    String sql = 'DELETE FROM users WHERE id = ?';
    try {
      final db = await _db;
      int result = await db.rawDelete(sql, [id]);
      if (result > 0) {
        print('✅ 회원 삭제 완료: $id');
        return true;
      }
    } catch (e) {
      print('❌ 회원 삭제 실패: $e');
    }
    return false;
  }
}
